Fix utility meter reset without tarrifs

gibwar opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
When using a utility_meter helper set up with a "Meter reset cycle" of "No cycle" it is impossible to properly reset the meter to 0 and utilize the "last reset" and "last period" attributes on the helper.

With these changes the service call can now directly reset a meter in such a configuration and the UI selector has been updated to include utility meters in the selection list.

## Type of change
- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #96920
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

gibwar committed:
Fix utility meter reset without tarrifs

When using a utility_meter helper set up with a "Meter reset cycle" of
"No cycle" it is impossible to properly reset the meter to 0 and utilize
the "last reset" and "last period" attributes on the helper.

With these changes the service call can now directly reset a meter in
such a configuration and the UI selector has been updated to include
utility meters in the selection list.

Fixes #96920

dgomes commented on homeassistant/components/utility_meter/services.yaml:
```diff
@@ -3,7 +3,9 @@
-      domain: select
+      - domain: select
+      - domain: sensor
+        integration: utility_meter
```
This creates an UX bug where users now expect to be able to reset individual tariff sensors...

gibwar replied:
I can certainly revert it to enable this functionality in YAML mode only. I don't use the tariff feature myself but is there a problem if they reset a tariff sensor directly?

dgomes replied:
if will not work

dgomes replied:
It's a UX aspect, reset service is used to reset all sensors at once (target is select), with this PR users can reset a single sensor and miss the remaining ones.

Why don't you setup a **single** tariff ? this way you could use reset without any change...

gibwar replied:
Looking at the [documentation](https://www.home-assistant.io/integrations/utility_meter/) I already am running in single tariff mode.

> ### Supported tariffs
> 
> A list of supported tariffs, leave empty if only a single tariff is needed.

The utility meter is perfect for my needs except for being able to manually reset it.

gibwar committed:
Undo UI target selection modes

As requested to not cause confusion by being able to reset individual
meters when using multiple tariffs.

gladhorn reviewed (approved):
Looks sensible to me. Thanks for adding to the test!

gibwar commented:
Rebased on to the latest dev

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -557,7 +557,7 @@ async def _async_reset_meter(self, event):
     async def async_reset_meter(self, entity_id):
         """Reset meter."""
-        if self._tariff_entity != entity_id:
+        if entity_id not in {self._tariff_entity, self.entity_id}:
```
```suggestion
        if self._tariff != None and self._tariff_entity != entity_id:
```

dgomes replied:
This way we don't change the current behaviour in which utility_meters with multiple tariffs are only reset through the `select.*` entity.

with your proposed change, you would enable individual tariff resets which would be an unintended change (IMHO)

gibwar replied:
I see what you're saying, I'll test it out locally and push up an updated branch here soon.

gibwar replied:
Tested and updated!

gibwar committed:
Adjust logic to only reset standalone sensors

The original logic would allow resetting meters with multiple tariffs
which was an unintended side effect.

gibwar commented:
Applied requested changes and updated to the latest dev build.

gibwar committed:
Merge branch 'dev' into fork-dev

pmannk commented:
Looking forward to this merge. I have a convoluted setup of sensors and automations trying to work around the current utility meter behaviour!

gibwar commented:
Anything more I need to do to get this merged (hopefully before 2024.3)? It looks like it's ready to merge but I don't have permission to do so. I see the following:

> **This branch has no conflicts with the base branch**
> Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.

dgomes commented:
It awaits at least 1 more core developer review

gibwar commented:
Fantastic! I wanted to make sure I wasn't missing something on my end. Thanks!

emontnemery reviewed (approved):
Thanks, @gibwar 👍
